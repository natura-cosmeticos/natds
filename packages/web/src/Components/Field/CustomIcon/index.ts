import CustomIcon from './CustomIcon'

export { ICustomIconProps } from './CustomIcon.props'
export { CustomIcon }
export default CustomIcon
