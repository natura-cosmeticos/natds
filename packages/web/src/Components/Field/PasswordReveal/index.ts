import PasswordReveal from './PasswordReveal'

export { IPasswordRevealProps } from './PasswordReveal.props'
export { PasswordReveal }
export default PasswordReveal
