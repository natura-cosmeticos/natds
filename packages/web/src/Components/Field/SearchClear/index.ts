import SearchClear from './SearchClear'

export { ISearchClearProps } from './SearchClear.props'
export { SearchClear }
export default SearchClear
