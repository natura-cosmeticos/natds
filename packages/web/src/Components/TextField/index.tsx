import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { tokens } from '@naturacosmeticos/natds-styles';
import { withTheme } from '@material-ui/styles';
import ErrorIcon from '@material-ui/icons/HighlightOffOutlined';
import SuccessIcon from '@material-ui/icons/CheckCircleOutline';

import { ITextFieldProps, getProp, getColorByState, stateStyles } from './shared';

import Field from './Field';

const TextField: FunctionComponent<ITextFieldProps> = (props: ITextFieldProps) => {
  const {
    id,
    label,
    helpText,
    theme,
    required = false,
    disabled = false,
    state
  } = props;

  const content = required ? `${label} *` : label;
  const IconState = stateIcons[String(state)];
  const stateIcon = IconState && <IconState theme={theme} />;

  return (
    <Container theme={theme}>
      <Label
        theme={theme}
        htmlFor={id}
        state={state}
        disabled={disabled}>
        {content}
      </Label>
      <Field {...props} />
      <HelpText
        theme={theme}
        state={state}
        disabled={disabled}>
        {stateIcon}
        {helpText}
      </HelpText>
    </Container>
  );
};

export default withTheme(TextField);

const Container = styled.div`
  display: flex;
  flex-flow: column nowrap;
  font-family: ${getProp('typography', 'fontFamily')};
`;

const Label = styled.label`
  font-size: ${getProp('typography', 'subtitle2', 'fontSize')};
  font-weight: ${getProp('typography', 'subtitle2', 'fontWeight')};
  color: ${getColorByState(stateStyles.hover)};
  line-height: 1.2;
  padding: 0 0 ${tokens.spacing.spacingMicro}px;
  font-family: ${getProp('typography', 'subtitle2', 'fontFamily')}
`;

const HelpText = styled.span`
  font-size: ${getProp('typography', 'caption', 'fontSize')};
  font-weight: ${getProp('typography', 'caption', 'fontWeight')};
  color: ${getColorByState(stateStyles.hover)};
  line-height: 1.2;
  padding: ${tokens.spacing.spacingMicro}px 0 0;
  display: flex;
  align-items: center;
  font-family: ${getProp('typography', 'subtitle2', 'fontFamily')}
`;

const baseIcon = `
  width: 16px!important;
  height: 16px!important;
  margin-right: 4px;
`;

const IconError = styled(ErrorIcon)`${baseIcon}`;

const IconSuccess = styled(SuccessIcon)`${baseIcon}`;

const stateIcons = {
  'error': IconError,
  'success': IconSuccess,
};
