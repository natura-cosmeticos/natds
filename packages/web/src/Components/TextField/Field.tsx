import React, { FunctionComponent, useState } from 'react';
import styled from 'styled-components';
import MaskedInput from 'react-text-mask';
import { tokens } from '@naturacosmeticos/natds-styles';

import { ITextFieldProps, getProp, getBorderByState, stateStyles, getColorByState } from './shared';
import PasswordReveal from './PasswordReveal';
import SearchClear from './SearchClear';
import CustomIcon from './CustomIcon';

export const TEXT_TYPE = 'text';
export const SEARCH_TYPE = 'search';
export const PASSWORD_TYPE = 'password';

const Field: FunctionComponent<ITextFieldProps> = (props: ITextFieldProps) => {
  const {
    id,
    theme,
    required = false,
    disabled = false,
    state,
    multiline,
    mask,
    placeholder = '&nbsp;',
    type,
    onIconPress,
    icon,
    searchIcon,
    showPasswordIcon,
    hidePasswordIcon,
    ...rest
  } = props;

  const fieldType = multiline ? 'textarea' : (mask ? MaskedInput : 'input');
  const [showing, togglePasswordReveal] = useState(false);
  const [value, setValue] = useState("");
  const customType = showing ? TEXT_TYPE : type;
  const showPasswordReveal = type === PASSWORD_TYPE && !icon;
  const showSearchClear = type === SEARCH_TYPE && !icon;
  const hasIcon = !!showPasswordReveal || !!showSearchClear || !!icon;

  return (
    <FieldContainer theme={theme} disabled={disabled}>
      <FieldComponent
        theme={theme}
        id={id}
        type={customType || TEXT_TYPE}
        placeholder={placeholder}
        state={state}
        disabled={disabled}
        required={required}
        as={fieldType}
        value={value}
        hasIcon={hasIcon}
        mask={mask}
        onChange={(event: any) => setValue(event.target.value)}
        {...rest}
      />
      {showPasswordReveal && <PasswordReveal
        theme={theme}
        showing={showing}
        onTogglePasswordReveal={togglePasswordReveal}
        showPasswordIcon={showPasswordIcon}
        hidePasswordIcon={hidePasswordIcon}
      />}
      {showSearchClear && <SearchClear
        theme={theme}
        onClearSearch={setValue}
        searchIcon={searchIcon}
      />}
      {icon && <CustomIcon
        theme={theme}
        icon={icon}
        onIconPress={onIconPress}
      />}
    </FieldContainer>
  );
};

export default Field;

const IconPad = `${tokens.spacing.spacingSmall}px ${tokens.spacing.spacingLarge}px ${tokens.spacing.spacingSmall}px ${tokens.spacing.spacingSmall}px`;

const FieldContainer = styled.div`
  display: flex;
  flex-flow: column nowrap;
  position: relative;
  fill: ${getColorByState(stateStyles.filled)};
`;

export const FieldComponent = styled.input`
  background-color: transparent;
  border: 0;
  border-radius: ${getProp('shape', 'borderRadius')}px;
  box-sizing: border-box;
  width: 100%;
  line-height: ${tokens.spacing.spacingStandard}px;
  min-height: ${tokens.spacing.spacingLarge}px;
  outline: none;
  color: ${getProp('palette', 'text', 'primary')};
  font-size: ${getProp('typography', 'body2', 'fontSize')};
  font-weight: ${getProp('typography', 'body2', 'fontWeight')};
  font-family: ${getProp('typography', 'fontFamily')};
  flex: 1 1 100%;
  padding: ${({ hasIcon }) => hasIcon ? IconPad : `${tokens.spacing.spacingSmall}px`};
  resize: vertical;

  &:disabled,
  &:disabled:hover {
    color: ${getProp('palette', 'text', 'hint')};
    box-shadow: ${getProp('palette', 'text', 'hint')} 0 0 0 1px;
  }

  &::placeholder {
    color: ${getProp('palette', 'text', 'secondary')};
  }

  &:disabled::placeholder {
    color: ${getProp('palette', 'text', 'hint')};
  }

  &:placeholder-shown {
    box-shadow: ${getBorderByState(stateStyles.default)};
  }

  &:not(:placeholder-shown) {
    box-shadow: ${getBorderByState(stateStyles.filled)};
  }

  &:focus:not(:read-only) {
    box-shadow: ${getBorderByState(stateStyles.focus)};
  }

  &:hover:not(:read-only):not(:disabled):not(:focus) {
    box-shadow: ${getProp('palette', 'text', 'secondary')} 0 0 0 1px;
  }
`;
