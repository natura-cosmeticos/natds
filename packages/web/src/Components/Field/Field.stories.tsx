import { Meta, Story } from '@storybook/react'
import * as React from 'react'
import { Field } from './Field'
import { ITextFieldProps } from '../TextField'

export default {
  component: Field,
  title: 'Utilities/Field'
} as Meta

const Template : Story<ITextFieldProps> = (args: ITextFieldProps) => (
  <Field id={args.id} {...args} />
)

export const Playground : Story<ITextFieldProps> = Template.bind({})
