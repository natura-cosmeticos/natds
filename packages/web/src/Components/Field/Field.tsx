/* eslint-disable @typescript-eslint/ban-ts-comment,@typescript-eslint/no-explicit-any */
import * as React from "react";
import MaskedInput from "react-text-mask";
import PasswordReveal from "./PasswordReveal";
import SearchClear from "./SearchClear";
import CustomIcon from "./CustomIcon";
import { ITextFieldProps } from "../TextField";
import { IInputProps, Input } from "../Input";

export const TEXT_TYPE = "text";
export const SEARCH_TYPE = "search";
export const PASSWORD_TYPE = "password";

export const Field = React.forwardRef<HTMLInputElement | HTMLTextAreaElement, ITextFieldProps>(
  // eslint-disable-next-line complexity,max-statements
  (props: ITextFieldProps, ref) => {
    const {
      className,
      disabled = false,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      helpText,
      hidePasswordIcon,
      icon,
      id,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      label,
      mask,
      maxLength,
      multiline,
      onChange,
      onIconPress,
      placeholder = "",
      required = false,
      searchIcon,
      showPasswordIcon,
      state,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      theme: providedTheme,
      type,
      ...otherProps
    } = props;

    let fieldType : IInputProps["inputComponent"] = "input";

    if (multiline) {
      fieldType = "textarea";
    } else if (mask) {
      fieldType = MaskedInput;
    }

    const [showing, togglePasswordReveal] = React.useState(false);
    const [value, setValue] = React.useState("");
    const customType = showing ? TEXT_TYPE : type;
    const showPasswordReveal = type === PASSWORD_TYPE && !icon;
    const showSearchClear = type === SEARCH_TYPE && !icon;
    const hasIcon = Boolean(showPasswordReveal) || Boolean(showSearchClear) || Boolean(icon);

    const handleChange = (event: { target: { value: React.SetStateAction<string>; }; }) => {
      setValue(event.target.value);
      if (onChange) {
        onChange(event);
      }
    };

    const clearSearch = (event: any) => {
      setValue("");
      if (onChange) {
        onChange(event);
      }
    };

    let endAdornment : React.ReactNode = null;

    if (showPasswordReveal) {
      endAdornment = <PasswordReveal
        hidePasswordIcon={hidePasswordIcon}
        onTogglePasswordReveal={togglePasswordReveal}
        showing={showing}
        showPasswordIcon={showPasswordIcon}
      />;
    }
    if (showSearchClear) {
      endAdornment = <SearchClear
        onClearSearch={clearSearch}
        searchIcon={searchIcon}
      />;
    }
    if (icon) {
      // @ts-ignore
      endAdornment = <CustomIcon icon={icon} onIconPress={onIconPress} />;
    }

    return (
      <Input
        disabled={disabled}
        endAdornment={endAdornment}
        hasIcon={hasIcon}
        id={id}
        inputComponent={fieldType}
        inputProps={{
          className: className && `${className}__input`,
          mask,
          maxLength: maxLength
        }}
        multiline={multiline}
        onChange={ handleChange }
        placeholder={placeholder}
        ref={ref}
        required={required}
        state={state}
        type={customType || TEXT_TYPE}
        value={value}
        disableUnderline={true}
        {...otherProps}
      />
    );
  });

Field.displayName = "Field";

export default Field;
