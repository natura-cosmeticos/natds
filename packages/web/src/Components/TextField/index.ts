import TextField from './TextField'

export { ITextFieldProps } from './TextField.props'
export { TextField }
export default TextField
