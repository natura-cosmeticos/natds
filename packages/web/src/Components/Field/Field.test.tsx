import * as React from 'react'
import * as TestRenderer from 'react-test-renderer'
import { Field } from './Field'
import { ITextFieldProps } from '../TextField'
import { themes } from '../../Themes'

const getProps = (props: Partial<ITextFieldProps> = {}) => {
  const {
    disabled,
    helpText = 'Assistive text',
    id = 'field',
    label = 'Label',
    mask,
    maxLength,
    placeholder = 'Placeholder',
    required,
    state,
    theme = themes.natura.light,
    type
  } = props

  return {
    disabled,
    helpText,
    id,
    label,
    mask,
    maxLength,
    placeholder,
    required,
    state,
    theme,
    type
  }
}

describe('Field component', () => {
  it('should match snapshot', () => {
    const testRenderer: TestRenderer.ReactTestRenderer = TestRenderer.create(<Field id="test-field" />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render single line input', () => {
    const props = getProps()
    const testRenderer = TestRenderer.create(<Field {...props} />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render single line masked input', () => {
    const props = getProps({ mask: ['v'] })
    const testRenderer = TestRenderer.create(<Field {...props} />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render input for password types', () => {
    const props = getProps({ type: 'password', value: 'new password' })
    const testRenderer = TestRenderer.create(<Field {...props} />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render input for search type', () => {
    const props = getProps({ type: 'search', value: 'new search' })
    const testRenderer = TestRenderer.create(<Field {...props} />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render input for search types and onChange', () => {
    const props = getProps({ type: 'search' })
    const onChange = jest.fn()
    const value = 'new search'
    const testRenderer = TestRenderer.create(<Field {...props} onChange={onChange} value={value} />)

    expect(testRenderer).toMatchSnapshot()
  })

  it('should render input with maxLength attribute equal 3', () => {
    const props = getProps({ maxLength: '3' })
    const testRenderer = TestRenderer.create(<Field {...props} />)

    expect(testRenderer).toMatchSnapshot()
  })
})
