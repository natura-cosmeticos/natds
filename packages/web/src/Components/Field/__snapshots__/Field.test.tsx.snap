// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field component should match snapshot 1`] = `
<div
  className="MuiInputBase-root MuiInput-root"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input"
    disabled={false}
    id="test-field"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    required={false}
    type="text"
  />
</div>
`;

exports[`Field component should render input for password types 1`] = `
<div
  className="MuiInputBase-root MuiInput-root MuiInputBase-adornedEnd"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
    disabled={false}
    id="field"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="password"
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root NatDSPasswordReveal-root-1"
    focusable="false"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M12 6c3.79 0 7.17 2.13 8.82 5.5C19.17 14.87 15.79 17 12 17s-7.17-2.13-8.82-5.5C4.83 8.13 8.21 6 12 6m0-2C7 4 2.73 7.11 1 11.5 2.73 15.89 7 19 12 19s9.27-3.11 11-7.5C21.27 7.11 17 4 12 4zm0 5c1.38 0 2.5 1.12 2.5 2.5S13.38 14 12 14s-2.5-1.12-2.5-2.5S10.62 9 12 9m0-2c-2.48 0-4.5 2.02-4.5 4.5S9.52 16 12 16s4.5-2.02 4.5-4.5S14.48 7 12 7z"
    />
  </svg>
</div>
`;

exports[`Field component should render input for search type 1`] = `
<div
  className="MuiInputBase-root MuiInput-root MuiInputBase-adornedEnd"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch"
    disabled={false}
    id="field"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="search"
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root NatDSSearchClear-root-2"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M18.3 5.71a.9959.9959 0 00-1.41 0L12 10.59 7.11 5.7a.9959.9959 0 00-1.41 0c-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
    />
  </svg>
</div>
`;

exports[`Field component should render input for search types and onChange 1`] = `
<div
  className="MuiInputBase-root MuiInput-root MuiInputBase-adornedEnd"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiInputBase-inputTypeSearch MuiInput-inputTypeSearch"
    disabled={false}
    id="field"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="search"
    value="new search"
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root NatDSSearchClear-root-2"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M18.3 5.71a.9959.9959 0 00-1.41 0L12 10.59 7.11 5.7a.9959.9959 0 00-1.41 0c-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
    />
  </svg>
</div>
`;

exports[`Field component should render input with maxLength attribute equal 3 1`] = `
<div
  className="MuiInputBase-root MuiInput-root"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input"
    disabled={false}
    id="field"
    maxLength="3"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="text"
  />
</div>
`;

exports[`Field component should render single line input 1`] = `
<div
  className="MuiInputBase-root MuiInput-root"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input"
    disabled={false}
    id="field"
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="text"
  />
</div>
`;

exports[`Field component should render single line masked input 1`] = `
<div
  className="MuiInputBase-root MuiInput-root"
  onClick={[Function]}
  required={false}
  style={
    Object {
      "marginTop": 0,
    }
  }
>
  <input
    className="MuiInputBase-input MuiInput-input"
    disabled={false}
    id="field"
    inputRef={[Function]}
    mask={
      Array [
        "v",
      ]
    }
    onAnimationStart={[Function]}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Placeholder"
    required={false}
    type="text"
  />
</div>
`;
